test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -0,0 +1,223 @@
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/integration/framework"
+	testutils "k8s.io/kubernetes/test/utils"
+	"sigs.k8s.io/yaml"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+var (
+	configFile = "config/performance-config.yaml"
+)
+
+// TestCase configures a test case to run the scheduler performance test. Users should be able to
+// provide this via a YAML file.
+//
+// It specifies nodes and pods in the cluster before running the test. It also specifies the pods to
+// schedule during the test. The config can be as simple as just specify number of nodes/pods, where
+// default spec will be applied. It also allows the user to specify a pod spec template for more compicated
+// test cases.
+//
+// It also specifies the metrics to be collected after the test. If nothing is specified, default metrics
+// such as scheduling throughput and latencies will be collected.
+type TestCase struct {
+	// description of the test case
+	Desc string
+	// configures nodes in the cluster
+	Nodes nodeCase
+	// configures pods in the cluster before running the tests
+	InitPods podCase
+	// pods to be scheduled during the test.
+	PodsToSchedule podCase
+	// optional, feature gates to set before running the test
+	FeatureGates map[featuregate.Feature]bool
+}
+
+type nodeCase struct {
+	Num                      int
+	NodeTemplatePath         *string
+	NodeAllocatableStrategy  *testutils.NodeAllocatableStrategy
+	LabelNodePrepareStrategy *testutils.LabelNodePrepareStrategy
+}
+
+type podCase struct {
+	Num                               int
+	PodTemplatePath                   *string
+	ReplicationControllerName         *string
+	PersistentVolumeTemplatePath      *string
+	PersistentVolumeClaimTemplatePath *string
+}
+
+func TestSchedulerPerformance(t *testing.T) {
+	var tests []TestCase
+	getSpec(&configFile, &tests)
+
+	for _, test := range tests {
+		name := fmt.Sprintf("%v/%vNodes/%vPods", test.Desc, test.Nodes.Num, test.InitPods.Num)
+		t.Run(name, func(t *testing.T) {
+			for feature, flag := range test.FeatureGates {
+				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, feature, flag)()
+			}
+
+			var nodeStrategy testutils.PrepareNodeStrategy
+			nodeStrategy = &testutils.TrivialNodePrepareStrategy{}
+
+			if test.Nodes.NodeAllocatableStrategy != nil {
+				nodeStrategy = test.Nodes.NodeAllocatableStrategy
+			} else if test.Nodes.LabelNodePrepareStrategy != nil {
+				nodeStrategy = test.Nodes.LabelNodePrepareStrategy
+			}
+
+			setupPodStrategy := getPodStrategy(test.InitPods)
+			testPodStrategy := getPodStrategy(test.PodsToSchedule)
+
+			var nodeSpec *v1.Node
+			if test.Nodes.NodeTemplatePath != nil {
+				nodeSpec = &v1.Node{}
+				getSpec(test.Nodes.NodeTemplatePath, nodeSpec)
+			}
+
+			perfScheduling(test.Nodes.Num, test.InitPods.Num, test.PodsToSchedule.Num, nodeStrategy, setupPodStrategy, testPodStrategy, nodeSpec)
+		})
+	}
+}
+
+func perfScheduling(numNodes, numExistingPods, minPods int,
```

Comment:
why don't we just pass TestCase as argument

Reply:
Copied from the old benchmark test, also it will be more convenient to use  `numNodes` than `testCase.Nodes.num` inside this function

Reply:
I found it weird that we already have a struct that contains all the arguments to this func, and we still decompose it. Also passing TestCase directly makes this more extensible (e.g., add a new parameter to TestCase in the future)

Reply:
I imagine `perfScheduling` will be the generic function to invoke tests, and it should take the generic `TestCase` argument. You can always create helper functions such as `simplePerfScheduling` with simplified arguments. But eventually I assume those simplified helper functions will call this one.

Reply:
Currently, the `TestCase` struct doesn't contain the `nodeStrategy`、`setupPodStrategy`、`testPodStrategy` object, shall we move the above logic into `perfScheduling`?

Reply:
Sounds good to me. 

Alternatively you can "decorate" `TestCase` by adding those strategies before passing to `perfScheduling`.